students/brandon_nguyen/session05/mailroom3.py

```diff
@@ -0,0 +1,168 @@
+                     " Average Gift")
+    line = '-'*len(header_string)
+    print(header_string)
+    print(line)
+    for donor, total, num, avg in newList:
+        print(" {:<18}     $  {:>12.2f} {:>6} {:>6} {:>8.2f}".
+              format(donor, total[0], num[0], "$", avg[0]))
+    print("\n\n")
+
+
+def sort_sum():
+    # creating a new list with computed value for easy printing.
+    """
+    This function returned a sorted list of by on order amount3
+    """
+    dblist = []
+    dbSum = []
+    # quick way to convert back to list to reuse existing code.
+    # list comprehension
+    [dblist.append((i, j)) for i, j in donor_db.items()]
+    [dbSum.append((name, [round(sum(donation), 2)], [len(donation)],
+     [round(sum(donation)/len(donation), 2)]),) for name, donation in dblist]
+    return sorted(dbSum, key=lambda donor: donor[1], reverse=True)
+
+
+def list_donors():
+    """
+    Basic Name of donor return in Order by First Name
+    """
+    lst = []
+    print()
+    [lst.append(person) for person in donor_db]  # list comprehension converted
+    lst.sort()
+    for name in lst:
+        print(name)
+    print()
+
+
+def update_donation():
+        input_person = input("\nPlease enter donor full names: ")
+        while True:
+            try:
+                input_donation = float(input("Please enter donation amount for " +
+                                       input_person + ":>> "))
+                break
+            except ValueError:
+                print("\nPlease Reenter amount in float.")
+
+        # existing people add to donation
+        if input_person in donor_db:
+            new_Donation = donor_db.get(input_person)
+            new_Donation.append(input_donation)
+            donor_db[input_person] = new_Donation
+        else:
+            donor_db.update({input_person: [input_donation]}) 
+        print()
+        print(email_template(input_person, input_donation))
+
+
+def email_template(name, newAmout):
+    e_frmt = {
+                  0: 'Dear',
+                  1: 'Thank you for your kind donation of ',
+                  2: 'It will be put to very good use',
+                  3: 'Sincerely, ',
+                  4: '-The Team'
+                  }
+    # Not the best way to do this.
+    txt = ("\n{} {},\n\n {:>45}${:.2f}.\n\n {: >40}.\n\n{:>40}\n"
+           "{:>42}".format(e_frmt.get(0), name, e_frmt.get(1), newAmout,
+                           e_frmt.get(2), e_frmt.get(3), e_frmt.get(4)))
+    return txt  # this way we can print to file
+
+
+def send_ty_all():
+    for personName in donor_db:
+        # assuming that the last donation appended to last
+        letter = email_template(personName, donor_db[personName][-1])
+        textfile = (personName.replace(" ", "_") + "_" +
+                    str(datetime.now()).replace(" ", "_")+".txt")
+        with open(textfile, 'w') as file_object:
+            file_object.write(letter)
```

Comment:
A good place for a `try`-`except` for cases when the file writing fails due to issues like `PermissionError` or `IOError`...